Transifex implementation

sgiehl opened the pull request:
As part of #7567 we need to switch our console commands to use Transifex.com API instead of oTrance.
The new implementation is a lot slower than before as we currently have a resource per language file (plugin). So we have 57 resources and 66 languages. The current implementation sends around 3700 api requests to transifex to update each resource and language as I didn't find a way to load "bulk data" eg. all translations over all resources and/or languages.
I've asked the tech support from transifex if there is any possibility to do that.
Let's see what they answer, otherwise the update will have a duration of about 10 minutes. (Which would be ok, I guess, as we are not doing it that often)

sgiehl committed:
refs #7567 - use transifex API to fetch translations

sgiehl committed:
refs #7567 - improved usage of transifex api

mnapoli commented:
Is it possible to do the "first translation update" in a separate pull request or would it be too much of a hassle? That would allow to review the diff (when the PR is ready for review), right now there are 600 files changed and the diff is too huge to load.

sgiehl committed:
removed now useless CompareKeys command; updated doc blocks and comments

sgiehl commented:
Good point. I've removed the commit.

sgiehl committed:
refs #7567 - use transifex statistics api and git modification time to determine which resources needs to be updated

sgiehl committed:
small fixes

sgiehl commented:
Good for review. As soon as it was merged, I'll create a first automatic translation update pull request using the new transifex api

mnapoli commented on core/Translation/Transifex/API.php:
```diff
@@ -0,0 +1,143 @@
+        curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
+        $response = curl_exec($curl);
+        $httpStatus = curl_getinfo($curl, CURLINFO_HTTP_CODE);
+        curl_close($curl);
```
Any reason not to use the `Http` class?

sgiehl replied:
No. I'll change that.

mnapoli commented on core/Translation/Transifex/API.php:
```diff
@@ -0,0 +1,143 @@
+     */
+    public function getAvailableResources()
+    {
+        static $resources;
```
Those should be class properties, static means they will be shared across instances

sgiehl replied:
It was on purpose to share that, as otherwise those data would be refetched for each command executed. That saves a lot of API requests.

mnapoli replied:
OK I see, but those resources depend on the state of the class. If I create a new instance of the class with a different Transifex project this is broken. I agree that this might never happen though, but using the cache system might be a better option as it's what we want to do (cache the results)? And probably not so much complicated (I don't really know I've never used it yet)?

sgiehl replied:
Hm. Using the cache might be an Option aswell. I'll check that tomorrow.

sgiehl replied:
I've relaced it with cache usage. See 93667de

mnapoli replied:
Awesome!

mnapoli commented on plugins/LanguagesManager/Commands/FetchFromTransifex.php:
```diff
@@ -0,0 +1,114 @@
+
+/**
+ */
+class FetchFromTransifex extends TranslationBase
```
Since the command is `translations:fetch`, maybe it would make more sense to keep the class name generic too? (e.g. `UpdateTranslations`?) That way it would avoid the situation here where `FetchFromOtrance` is deleted and `FetchFromTransifex` is created whenever we change again from provider?

sgiehl replied:
Good Point

mnapoli commented on plugins/LanguagesManager/Commands/Update.php:
```diff
@@ -55,79 +48,78 @@ protected function execute(InputInterface $input, OutputInterface $output)
 
-        $progress->finish();
-        $output->writeln("Finished.");
+        $output->writeln("Finished in " . round(microtime(true)-$start, 3) . "s");
```
FYI there is `Piwik\Timer` that can do that (don't need to change that code, it's just FYI in case you didn't know about it)

sgiehl replied:
Didn't know...

mnapoli commented:
Looks good to me, I've commented minor details. Just out of curiosity why are "fetch translations" and "update translations" separate operations?

mattab commented:
Nice progress @sgiehl 
 
I'm curious: did you hear from Transifex support about possible speed improvements to API imports?

sgiehl committed:
small improvements

sgiehl commented:
@mnapoli I don't know the reason why I have splitted that in the first place. Maybe as it was easier to build that part by part.
I've changed some of the stuff you commented and rebased on master. I'll merge it now and create a new pull request for translation updates using the new stuff :)
